pkg/daemon/pinned_image_set_manager.go

```diff
@@ -0,0 +1,1096 @@
+
+	mcpLister mcfglistersv1.MachineConfigPoolLister
+	mcpSynced cache.InformerSynced
+
+	mcfgClient mcfgclientset.Interface
+
+	prefetchCh chan prefetch
+
+	criClient *cri.Client
+
+	// minimum storage available after prefetching
+	minStorageAvailableBytes resource.Quantity
+	// path to the authfile
+	authFilePath string
+	// endpoint of the container runtime service
+	runtimeEndpoint string
+	// timeout for the prefetch operation
+	prefetchTimeout time.Duration
+	// backoff configuration for retries.
+	backoff wait.Backoff
+
+	syncHandler              func(string) error
+	enqueueMachineConfigPool func(*mcfgv1.MachineConfigPool)
+	queue                    workqueue.RateLimitingInterface
+	featureGatesAccessor     featuregates.FeatureGateAccess
+
+	// mutex to protect the cancelFn
+	mu       sync.Mutex
+	cancelFn context.CancelFunc
+}
+
+// NewPinnedImageSetManager creates a new pinned image set manager.
+func NewPinnedImageSetManager(
+	nodeName string,
+	criClient *cri.Client,
+	mcfgClient mcfgclientset.Interface,
+	imageSetInformer mcfginformersv1alpha1.PinnedImageSetInformer,
+	nodeInformer coreinformersv1.NodeInformer,
+	mcpInformer mcfginformersv1.MachineConfigPoolInformer,
+	minStorageAvailableBytes resource.Quantity,
+	runtimeEndpoint string,
+	authFilePath string,
+	prefetchTimeout time.Duration,
+	featureGatesAccessor featuregates.FeatureGateAccess,
+) *PinnedImageSetManager {
+	p := &PinnedImageSetManager{
+		nodeName:                 nodeName,
+		mcfgClient:               mcfgClient,
+		runtimeEndpoint:          runtimeEndpoint,
+		authFilePath:             authFilePath,
+		prefetchTimeout:          prefetchTimeout,
+		minStorageAvailableBytes: minStorageAvailableBytes,
+		featureGatesAccessor:     featureGatesAccessor,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pinned-image-set-manager"),
+		prefetchCh:               make(chan prefetch, defaultPrefetchWorkers*2),
+		criClient:                criClient,
+		backoff: wait.Backoff{
+			Steps:    maxRetries,
+			Duration: retryDuration,
+			Factor:   retryFactor,
+			Cap:      retryCap,
+		},
+	}
+
+	imageSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addPinnedImageSet,
+		UpdateFunc: p.updatePinnedImageSet,
+		DeleteFunc: p.deletePinnedImageSet,
+	})
+
+	mcpInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addMachineConfigPool,
+		UpdateFunc: p.updateMachineConfigPool,
+		DeleteFunc: p.deleteMachineConfigPool,
+	})
+
+	p.syncHandler = p.sync
+	p.enqueueMachineConfigPool = p.enqueue
+
+	p.imageSetLister = imageSetInformer.Lister()
+	p.imageSetSynced = imageSetInformer.Informer().HasSynced
+
+	p.nodeLister = nodeInformer.Lister()
+	p.nodeListerSynced = nodeInformer.Informer().HasSynced
+
+	p.mcpLister = mcpInformer.Lister()
+	p.mcpSynced = mcpInformer.Informer().HasSynced
+
+	return p
+}
+
+func (p *PinnedImageSetManager) sync(key string) error {
```

Comment:
I guess my main question here is whether there's any coordination going on with the main MCD sync loop (apologies if there is and I'm just missing it).

Although not a super big concern, I'm mostly wondering if it's ever a problem if both the main MCD upgrade loop upgrades the node while this is trying to pull an image, and:
1. we think there's enough space here, but while we're pulling, the main MCD loop also pulls the new oscontainer, and combined they don't have enough space (maybe this would almost never happen so not really a concern?)
2. this pulls halfway and the daemon queues a reboot (presumably we'd start from scratch next boot, or maybe the daemon waits for this?)

Just wondering if there should by any coordination here

Reply:
good points so the design does take this into consideration , basically we do the following to ensure sanity.

> we think there's enough space here, but while we're pulling, the main MCD loop also pulls the new oscontainer, and combined they don't have enough space (maybe this would almost never happen so not really a concern?)

- before we pull any images we calculate the total payload size for all images.
- we don't pin the images until after all images are pulled this allows for an escape hatch for the kublet to GC if for some reason another large image pull happens .
- we currently require a 16GB buffer of extra space on top of our payload size.

This feels reasonable, kubelet can GC our images until we pin them.

> this pulls halfway and the daemon queues a reboot (presumably we'd start from scratch next boot, or maybe the daemon waits for this?)

No I think the signal that we are complete is the signal for any observer that upgrade is possible. In your example if the daemon does some work and that results in a reboot then we reconcile again on reboot.  There is no expectation that the images should be ready until the status reports complete.